### DISCUSSION TITLE
What is the right way to watch files from the language server.

### DISCUSSION THREAD

**Discussion by MechMel** [eclipse-langium/langium#901] [Q&A] [answered]
I'm making a programming language named Mynx. I want to watch all Mynx files in the workspace, and transpile them on edit. Here is how I am currently doing this. Is this the right way? Is there a better way to do this?
```
export function createMynxServices(context: DefaultSharedModuleContext): {
  shared: LangiumSharedServices;
  Mynx: MynxServices;
} {
  const shared = inject(
    createDefaultSharedModule(context),
    MynxGeneratedSharedModule,
  );
  const Mynx = inject(
    createDefaultModule({ shared }),
    MynxGeneratedModule,
    MynxModule,
  );
  /** Obviously this should go somewhere else in future. */
  shared.workspace.TextDocuments.onDidChangeContent((e) => {
    if (e.document.uri.endsWith(".mx")) {
      try {
        const langiumDocument =
          shared.workspace.LangiumDocuments.getOrCreateDocument(
            URI.parse(e.document.uri),
          );
        const model = langiumDocument.parseResult?.value as Model;
        generateJavaScript(
          model,
          URI.parse(e.document.uri).fsPath,
          "./builds/website",
        );
      } catch (e) {
        console.error(e);
      }
    }
  });
  shared.ServiceRegistry.register(Mynx);
  return { shared, Mynx };
}
```

**Comment by msujew** [comment, marked as answer]
Hey @MechMel,

While this correctly notifies you about file changes, the documents in question will not be in a good state to be used as a source of code generation. We go trough multiple async document builder phases, which are responsible for stuff like cross reference resolution and validation. Using `onDidChangeContent` doesn't allow you to wait until these phases have completed.

Instead, you can use the `DocumentBuilder.onBuildPhase(DocumentState.Validated, (documents, cancellationToken) => {})` callback. This has multiple advantages:

1. It only runs your callback once the validation has completed. This allows to check whether your document has any errors or any other issues which prevent it from successful code generation.
2. Cross reference resolution has also been performed on the document, so you don't need to worry about that as well
3. The callback gives you all documents which are affected by this change. For example, if you change a signature, another element in another file might become invalid. This will be caught by this callback.

Make sure to use the `cancellationToken` of the callback to cancel the generation once the user has typed something else into their editor, so you don't hog the resources of the language client unnecessarily.

**Reply by MechMel** [reply]
Sweet I figured you guys had a better way to do this. Works like a charm.
